### ISSUE TITLE
Recurse subdirectories for docker-entrypoint-initdb.d functionality

### ISSUE THREAD

**Issue by md5** [docker-library/postgres#179]
Over in https://github.com/appropriate/docker-postgis/issues/35, I had a user ask about adding their own script to `/docker-entrypoint-initdb.d` via `docker-compose` without overwriting the script I use for initializing PostGIS. My suggestion was to bind-mount individual scripts, but it dawned on me that it might be nice to be able to add scripts to subdirectories of `/docker-entrypoint-initdb.d` and have them sourced or executed according to the existing logic.

The main drawback I see of making this change is that it's possible that existing users of the `postgres` image may already have files in subdirectories of `/docker-entrypoint-initdb.d` that they don't want to have executed directly by the `postgres` entrypoint. Another minor but solvable issue is maintaining the sort order that scripts are executed in that the current globbing implicitly provides, which I believe I documented when this functionality was originally added:

> These initialization files will be executed in sorted name order as defined by the current locale, which defaults to `en_US.utf8`

**Comment by yosifkit** [comment]
I agree about the main drawback;  I'd rather not publish a breaking change.  The sorting of a `find` should be about the same as globing; we can even make sure that everything within a directory comes before sub-directories.  Maybe we hide it behind an environment flag to turn it on?  If we were to add it would we limit the depth when finding scripts/sql?

Maybe we just recommend users mount a script and have it run other scripts in their folders.

**Comment by md5** [comment]
@yosifkit I think your second suggestion sounds better than adding another environment variable

**Comment by md5** [comment]
Also, I thought `find` uses `readdir` order

**Comment by stepankuzmin** [comment]
My solution is mounting `./initdb.d` folder to `/initdb.d` and `./initdb.sh` script to `/docker-entrypoint-initdb.d/initdb.sh`:

```shell
for f in /initdb.d/*.sql; do
  echo "$0: running $f"
  "${psql[@]}" -f "$f"
done
```

**Comment by nullhack** [comment]
A simple solution is to create a file (`init.sh` inside `/docker-entrypoint-initdb.d/initdb.sh`) with the following content:

```bash
#!/usr/bin/env bash
find /docker-entrypoint-initdb.d -mindepth 2 -type f -print0 | while read -d $'\0' f; do
  case "$f" in
    *.sh)
      if [ -x "$f" ]; then
        echo "$0: running $f"
        "$f"
      else
        echo "$0: sourcing $f"
        . "$f"
      fi
      ;;
    *.sql)    echo "$0: running $f"; "${psql[@]}" -f "$f"; echo ;;
    *.sql.gz) echo "$0: running $f"; gunzip -c "$f" | "${psql[@]}"; echo ;;
    *)        echo "$0: ignoring $f" ;;
  esac
  echo
done
```

This will recurse all sub-directories inside `/docker-entrypoint-initdb.d/` and run the files the same way the original Dockerfile do

**Comment by alexander793** [comment]
> A simple solution is to create a file (`init.sh` inside `/docker-entrypoint-initdb.d/initdb.sh`) with the following content:
> 
> #!/usr/bin/env bash
> find /docker-entrypoint-initdb.d -mindepth 2 -type f -print0 | while read -d $'\0' f; do
>   case "$f" in
>     *.sh)
>       if [ -x "$f" ]; then
>         echo "$0: running $f"
>         "$f"
>       else
>         echo "$0: sourcing $f"
>         . "$f"
>       fi
>       ;;
>     *.sql)    echo "$0: running $f"; "${psql[@]}" -f "$f"; echo ;;
>     *.sql.gz) echo "$0: running $f"; gunzip -c "$f" | "${psql[@]}"; echo ;;
>     *)        echo "$0: ignoring $f" ;;
>   esac
>   echo
> done
> 
> This will recurse all sub-directories inside `/docker-entrypoint-initdb.d/` and run the files the same way the original Dockerfile do

In case anybody comes across this: The provided solution works, if your scripts do not depend on each other. If you need them to be executed in a specific order, prefix them with an increasing number and insert `| sort -z` to the second line in the provided script like: `find /docker-entrypoint-initdb.d -mindepth 2 -type f -print0 | sort -z | while read -d $'\0' f; do`.
This will sort all scripts in ascending order before they are executed.
